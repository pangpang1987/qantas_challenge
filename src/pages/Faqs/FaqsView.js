import React from "react";
import PropTypes from "prop-types";

import { ViewContainer, MainSection, Title, Body, RightPanel } from "./styled";
import PanelLink from "./PanelLink";

class FaqsView extends React.Component {
  state = {
    activeIndex: 0
  };
  componentDidMount() {
    this.props.actions.fetchFaqs();
  }

  setActiveIndex = activeIndex => {
    this.setState({ activeIndex });
  };

  render() {
    const { faqs, isLoading } = this.props;
    const { activeIndex } = this.state;

    if (isLoading) {
      return <div>Loading...</div>;
    }

    const activeFaq = faqs[activeIndex] || {};

    return (
      <ViewContainer>
        <MainSection>
          <Title>{activeFaq.title}</Title>
          <Body dangerouslySetInnerHTML={{ __html: activeFaq.body }} />
        </MainSection>
        <RightPanel>
          {faqs.map((faq, index) => (
            <PanelLink
              key={index}
              index={index}
              title={faq.title}
              activeIndex={activeIndex}
              setActiveIndex={this.setActiveIndex}
            />
          ))}
        </RightPanel>
      </ViewContainer>
    );
  }
}

FaqsView.propTypes = {
  faqs: PropTypes.array,
  isLoading: PropTypes.bool,
  actions: PropTypes.object
};

export default FaqsView;
