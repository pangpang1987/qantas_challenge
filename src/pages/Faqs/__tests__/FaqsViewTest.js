import React from "react";

import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import FaqsView from "../FaqsView";

describe("Component <FaqsView />", () => {
  const fetchFaqs = jest.fn();
  const actions = { fetchFaqs };
  const faqs = [
    { title: "test title here", body: "testBody" },
    { title: "element test", body: "<p>first</p><p>second</p>" }
  ];

  it("should render initial state properly", () => {
    const component = shallow(<FaqsView faqs={[]} isLoading={false} actions={actions} />);

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
    expect(fetchFaqs.mock.calls.length).toBe(1);
  });

  it("should render a loader when fetching data", () => {
    const component = shallow(<FaqsView faqs={[]} isLoading={true} actions={actions} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it("should render properly with data", () => {
    const component = shallow(<FaqsView faqs={faqs} isLoading={false} actions={actions} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it("should render properly when switch to another title", () => {
    const component = shallow(<FaqsView faqs={faqs} isLoading={false} actions={actions} />);
    component.setState({ activeIndex: 1 });
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
