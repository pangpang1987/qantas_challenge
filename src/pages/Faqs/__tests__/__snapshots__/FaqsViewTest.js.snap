// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <FaqsView /> should render a loader when fetching data 1`] = `
<div>
  Loading...
</div>
`;

exports[`Component <FaqsView /> should render initial state properly 1`] = `
<styled.div>
  <styled.section>
    <styled.h3 />
    <styled.div
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
  </styled.section>
  <styled.ul />
</styled.div>
`;

exports[`Component <FaqsView /> should render properly when switch to another title 1`] = `
<styled.div>
  <styled.section>
    <styled.h3>
      element test
    </styled.h3>
    <styled.div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>first</p><p>second</p>",
        }
      }
    />
  </styled.section>
  <styled.ul>
    <PanelLink
      activeIndex={1}
      index={0}
      key="0"
      setActiveIndex={[Function]}
      title="test title here"
    />
    <PanelLink
      activeIndex={1}
      index={1}
      key="1"
      setActiveIndex={[Function]}
      title="element test"
    />
  </styled.ul>
</styled.div>
`;

exports[`Component <FaqsView /> should render properly with data 1`] = `
<styled.div>
  <styled.section>
    <styled.h3>
      test title here
    </styled.h3>
    <styled.div
      dangerouslySetInnerHTML={
        Object {
          "__html": "testBody",
        }
      }
    />
  </styled.section>
  <styled.ul>
    <PanelLink
      activeIndex={0}
      index={0}
      key="0"
      setActiveIndex={[Function]}
      title="test title here"
    />
    <PanelLink
      activeIndex={0}
      index={1}
      key="1"
      setActiveIndex={[Function]}
      title="element test"
    />
  </styled.ul>
</styled.div>
`;
